consensus/istanbul/core/handler.go

```diff
@@ -172,43 +246,226 @@ func (c *core) handleMsg(payload []byte) error {
+		case istanbul.MsgCommit:
+			// Discard messages from previous views, unless they are commits from the previous sequence,
+			// with the same round as what we wound up finalizing, as we would be able to include those
+			// to create the ParentAggregatedSeal for our next proposal.
+			lastSubject, err := c.backend.LastSubject()
+			if err != nil {
+				return err
+			} else if view(m).Cmp(lastSubject.View) != 0 {
+				return errOldMessage
+			} else if lastSubject.View.Sequence.Cmp(common.Big0) == 0 {
+				// Don't handle commits for the genesis block, will cause underflows
+				return errOldMessage
+			}
+			return c.handleCheckedCommitForPreviousSequence(msg, msg.Commit())
+		case istanbul.MsgRoundChange:
+			// If the RC message is for the current sequence but a prior round, help the sender fast forward
+			// by sending back to it (not broadcasting) a round change message for our desired round.
+			if m.Height == c.current.Sequence().Uint64() {
+				logger.Trace("Sending round change for desired round to node with a previous desired round", "msg_round", m.Round)
+				c.sendRoundChangeAgain(msg.Address)
 			}
 		}
-		return err
+		return errOldMessage
 	}
 
+	// Check if the message is a future message and if so add it to the backlog
+	if IsFutureMsg(c.current.Sequence(), c.current.DesiredRound(), c.current.State(), view(m), msg.Code) {
+		// Store in backlog (if it's not from self)
+		if msg.Address != c.address {
+			c.backlog.store(msg)
+		}
+		return errFutureMessage
+	}
+
+	numberOfCommits := c.current.Commits().Size()
+	minQuorumSize := c.current.ValidatorSet().MinQuorumSize()
+	// At this point we know that messages are either for the current sequence
+	// and desired round or its a round change for the current sequence and
+	// desired or future round.
 	switch msg.Code {
 	case istanbul.MsgPreprepare:
-		return catchFutureMessages(c.handlePreprepare(msg))
+		preprepare := msg.Preprepare()
+		// Verify that the proposal is for the sequence number of the view we verified.
+		if m.Height != preprepare.Proposal.Number().Uint64() {
+			logger.Warn("Received preprepare with invalid block number")
+			return errInvalidProposal
+		}
+
+		// Check proposer is valid for the message's view (this may be a subsequent round)
+		headBlock, headProposer := c.backend.GetCurrentHeadBlockAndAuthor()
+		if headBlock == nil {
+			logger.Error("Could not determine head proposer")
+			return errNotFromProposer
+		}
+		proposerForMsgRound := c.selectProposer(c.current.ValidatorSet(), headProposer, m.Round)
+		if proposerForMsgRound.Address() != msg.Address {
+			logger.Warn("Ignore preprepare message from non-proposer", "actual_proposer", proposerForMsgRound.Address())
+			return errNotFromProposer
+		}
+
+		defer c.handlePrePrepareTimer.UpdateSince(time.Now())
+
+		if m.Round > 0 {
+			logger.Trace("Trying to move to round change certificate's round", "target round", m.Round)
+			err := c.startNewRound(new(big.Int).SetUint64(m.Round))
```

Comment:
Is this not an algorithm step?

Reply:
It definitely looks like it should be! Will look into this more, thanks